crates/ruff_server/src/trace.rs

```diff
@@ -0,0 +1,143 @@
+use lsp_types::TraceValue;
+use serde::Deserialize;
+use std::{
+    path::PathBuf,
+    sync::{Mutex, OnceLock},
+};
+use tracing::level_filters::LevelFilter;
+use tracing_subscriber::{fmt::time::Uptime, layer::SubscriberExt, Layer};
+
+use crate::server::ClientSender;
+
+const TRACE_ENV_KEY: &str = "RUFF_TRACE";
+
+static LOGGING_SENDER: OnceLock<ClientSender> = OnceLock::new();
+
+static TRACE_VALUE: Mutex<lsp_types::TraceValue> = Mutex::new(lsp_types::TraceValue::Off);
+
+pub(crate) fn set_trace_value(trace_value: TraceValue) {
+    let mut global_trace_value = TRACE_VALUE
+        .lock()
+        .expect("trace value mutex should be available");
+    *global_trace_value = trace_value;
+}
+
+pub(crate) fn init_tracing(
+    sender: ClientSender,
+    log_level: LogLevel,
+    log_file: Option<&std::path::Path>,
+) {
+    LOGGING_SENDER
+        .set(sender)
+        .expect("logging sender should only be initialized once");
+
+    let subscriber = tracing_subscriber::Registry::default().with(
+        tracing_subscriber::fmt::layer()
+            .with_timer(Uptime::default())
+            .with_thread_names(true)
+            .with_ansi(false)
+            .with_writer(TracingWriter::new(log_file))
+            .with_filter(TraceLevelFilter)
+            .with_filter(LogLevelFilter { filter: log_level }),
+    );
+
+    tracing::subscriber::set_global_default(subscriber)
+        .expect("should be able to set global default subscriber");
+}
+
+#[derive(Clone, Copy, Debug, Deserialize, Default, PartialEq, Eq, PartialOrd, Ord)]
+#[serde(rename_all = "lowercase")]
+pub(crate) enum LogLevel {
+    #[default]
+    Error,
+    Warn,
+    Info,
+    Debug,
+    Trace,
+}
+
+impl LogLevel {
+    fn trace_level(self) -> tracing::Level {
+        match self {
+            Self::Error => tracing::Level::ERROR,
+            Self::Warn => tracing::Level::WARN,
+            Self::Info => tracing::Level::INFO,
+            Self::Debug => tracing::Level::DEBUG,
+            Self::Trace => tracing::Level::TRACE,
+        }
+    }
+}
+
+struct LogLevelFilter {
+    filter: LogLevel,
+}
+
+struct TraceLevelFilter;
+
```

Comment:
What's the motivation of having two filters rather than just one?

Reply:
They're filtering by different things, and I'd rather keep them separated for modularity.

Reply:
Could you add a documentation explaining what they're filtering by then so that a future me understand the motivation of keeping them different and doesn't merge them together.